api/secret.go

```diff
@@ -316,5 +324,14 @@ func ParseSecret(r io.Reader) (*Secret, error) {
 		return nil, err
 	}
 
+	// If the data is null, add raw data to secret data if present
+	if secret.Data == nil {
+		data := make(map[string]interface{})
+		if err := jsonutil.DecodeJSONFromReader(&teebuf, &data); err != nil {
```

Comment:
I think you can do something like this:

```suggestion
		if err := mapstructure.Decode(secret, &data); err != nil {
```

and then we don't need a second copy of the stream above. I think this might be a good optimization because it prevents duplicating the stream every time we read, regardless of whether we have a `Data` field.  In this case, we'll only need to make a copy of the data on-demand.

How do you feel about this?

Reply:
Except for the fact that it doesn't work 😆 . I'd have to think a bit more about how to get this correct, but it seems like it should be achievable...

Reply:
I don't think this would work as the secret would not have any data after decode in this case

Reply:
What we're looking for would be a way to be aware that some JSON keys didn't map to the `Secret` upon decode. I'm pretty sure that isn't possible. I think in order to achieve a single read would be to forego the decode to a `Secret` and instead use custom processing. The second decode isn't ideal for performance reasons but it's likely only going to happen in very few cases. It's probably a reasonable tradeoff to avoid introducing potential regressions by changing the `vault read` decode semantics completely. If there is a low risk way to prevent a subsequent decode then I'm all for it though!

Reply:
Looks like you can just decode to a `map[string]interface{}` and then decode _that_ into `secret`/`secret.Data`.

What do you think about something like this instead?

```go
	// First decode the JSON into a map[string]interface{}
	data := make(map[string]interface{})
	if err := jsonutil.DecodeJSONFromReader(&buf, &data); err != nil {
		return nil, err
	}

	var secret Secret
	if err := mapstructure.Decode(data, &secret); err != nil {
		return nil, err
	}
	
	// If the data is null, add raw data to secret data if present
	if secret.Data == nil {
		if err := mapstructure.Decode(data, &secret.Data); err != nil {
			return nil, err
		}
	}
```

Reply:
I'm not sure if this gains us anything, but we at least avoid creating two copies of the buffer for every read.

Reply:
No worries, thanks for hopping in the thread! I'm actually sort of convinced that `mapstructure.Decode` with a decoder hook is the correct approach here: https://pkg.go.dev/github.com/mitchellh/mapstructure#DecodeHookFunc.

I think you might be able to use this avoid making a copy of the stream _and_ decode in one pass.

If we're not super concerned with optimizing this path, I'm cool with that. Just figured I'd raise the topic for conversation in case it was a concern.

Reply:
For full context, @akshya96 and I met synchronously about it last week so I had some prior context. It's totally worth exploring if there is a low risk of regression.